### ISSUE TITLE
ros2 support

### ISSUE THREAD

**Issue by Sollimann** [at-wat/mcl_3dl#391]
Hi,

First of all, thank you for your great work and contribution.
I was just wondering if there is planned for adding ros2 support?

All Best

**Comment by at-wat** [comment]
I don't have plan at now.
Contributions are very welcome! (but the design should be discussed before implementation)

**Comment by greymfm** [comment]
There exist (semi-)automated ROS1->ROS2 conversion tools:  https://github.com/awslabs/ros2-migration-tools
 Maybe useful...

**Comment by inkccc** [comment]
你好 请问现在有ros2版本的支持了吗

**Comment by Ryujiyasu** [comment]
Hi @at-wat, I'd like to work on porting mcl_3dl to ROS 2. Before starting implementation, I'd like to discuss the design as you suggested.

## Migration Plan

### Target
- **ROS 2 Jazzy** (latest LTS, Ubuntu 24.04)
- Approach: create a `ros2` branch based on `master`

### Build System
- `catkin` / `CMakeLists.txt` → `ament_cmake`
- `package.xml` format 2 → format 3 with `<build_type>ament_cmake</build_type>`

### Core API Migration
- `roscpp` → `rclcpp` (Node subclass or composition-based)
- `ros::Publisher` / `ros::Subscriber` → `rclcpp::Publisher` / `rclcpp::Subscription`
- `ros::Timer` → `rclcpp::TimerBase`
- `ros::NodeHandle` param access → `rclcpp::Node::declare_parameter` / `get_parameter`

### dynamic_reconfigure → ROS 2 Parameters
- `MCL3DLParams.cfg` (3 params: `std_warn_thresh_xy`, `std_warn_thresh_z`, `std_warn_thresh_yaw`) → `rclcpp` dynamic parameters with `on_set_parameters_callback`
- The cfg file itself will be removed

### Messages
- `mcl_3dl_msgs` (Status.msg) will also need to be ported: `message_generation` / `message_runtime` → `rosidl_default_generators` / `rosidl_default_runtime`
- I can submit a separate PR for `mcl_3dl_msgs` first

### TF2
- Already using `tf2_ros` (not the deprecated `tf`), so the migration should be straightforward
- `tf2_ros::TransformListener` / `Buffer` API is mostly compatible

### Launch
- `test.launch` (XML) → ROS 2 launch XML or Python launch file

### Tests
- `rostest` / `rosunit` → `launch_testing` / `ament_cmake_gtest`
- `roslint` → `ament_lint_auto` / `ament_cmake_clang_format` (or keep existing style)

### PCL
- `pcl_ros` dependency → `pcl_ros` (ROS 2 version exists in perception_pcl)
- PCL core usage should remain mostly unchanged

## Questions

1. **Branch strategy**: Should I target a `ros2` branch, or do you prefer a separate `ros2-jazzy` branch? Would you like to maintain ROS 1 and ROS 2 in the same branch (unlikely but asking)?
2. **Node composition**: Would you prefer a standalone executable node, or a composable node (component-based) that supports `rclcpp_components`? I'd suggest supporting both (composable + standalone wrapper), which is the modern ROS 2 convention.
3. **Launch format**: Preference for Python launch or XML launch for ROS 2?
4. **mcl_3dl_msgs**: Should I port this first as a prerequisite PR?

Looking forward to your feedback!

**Comment by at-wat** [comment]
In my opinion, the core logic should be separated to a pure C++ library first and a ROS 2 wrapper using the core logic library should be added, instead of directly converting the current ROS 1 code to ROS 2.

See https://github.com/at-wat/mcl_3dl/issues/433 for the details.
